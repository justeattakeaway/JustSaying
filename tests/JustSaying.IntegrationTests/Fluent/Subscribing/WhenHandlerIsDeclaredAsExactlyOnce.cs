using JustSaying.Messaging.MessageHandling;
using JustSaying.Messaging.Middleware;
using JustSaying.TestingFramework;
using Microsoft.Extensions.DependencyInjection;
using Newtonsoft.Json;

namespace JustSaying.IntegrationTests.Fluent.Subscribing;

public class WhenHandlerIsDeclaredAsExactlyOnce(ITestOutputHelper outputHelper) : IntegrationTestBase(outputHelper)
{
    [AwsFact]
    public async Task Then_The_Handler_Only_Receives_The_Message_Once()
    {
        // Arrange
        var messageLock = new MessageLockStore();
        var handler = new InspectableHandler<SimpleMessage>();

        var services = GivenJustSaying()
            .AddSingleton<IMessageLockAsync>(messageLock)
            .ConfigureJustSaying((builder) =>
                builder.WithLoopbackTopic<SimpleMessage>(UniqueName,
                    c => c.WithMiddlewareConfiguration(m =>
                        m.UseExactlyOnce<SimpleMessage>("lock-simple-message")
                            .UseDefaults<SimpleMessage>(handler.GetType()))))
            .AddJustSayingHandlers(new[] { handler });

        var message = new SimpleMessage();
        string json = "";
        await WhenAsync(
            services,
            async (publisher, listener, serviceProvider, cancellationToken) =>
            {
                await listener.StartAsync(cancellationToken);
                await publisher.StartAsync(cancellationToken);

                // Act
                await publisher.PublishAsync(message, cancellationToken);
                await publisher.PublishAsync(message, cancellationToken);

                dynamic middlewares = ((dynamic)listener.Interrogate().Data).Middleware;
                json = JsonConvert.SerializeObject(middlewares, Formatting.Indented)
                    .Replace(UniqueName, "TestQueueName");

                await Patiently.AssertThatAsync(() =>
                {
                    handler.ReceivedMessages.Where(m => m.Id.ToString() == message.UniqueKey())
                        .ShouldHaveSingleItem();
                });
            });

        json.ShouldMatchApproved(c => c.SubFolder("Approvals"));
    }
}