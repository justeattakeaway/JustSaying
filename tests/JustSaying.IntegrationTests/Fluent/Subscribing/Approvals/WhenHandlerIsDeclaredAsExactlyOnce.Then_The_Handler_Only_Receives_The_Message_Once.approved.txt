{
  "Middlewares": [
    {
      "MessageType": "SimpleMessage",
      "QueueName": "TestQueueName",
      "MiddlewareChain": [
        "ExactlyOnceMiddleware`1",
        "MessageContextAccessorMiddleware",
        "ErrorHandlerMiddleware",
        "LoggingMiddleware",
        "StopwatchMiddleware",
        "SqsPostProcessorMiddleware",
        "HandlerInvocationMiddleware`1"
      ]
    }
  ]
}